Refactor move user actions to 'gui.actions' namespace

gonzalo-bulnes opened the pull request:
# Description

:crystal_ball: **Review recommendations**: I split the changes in three commits, so that tests can be run (unchanged) after the class as been moved, then can be run again (without changes to the non-test code) after they are moved. :slightly_smiling_face: 

This is the 12th step to be split out of #1369.
It moves the `DeleteSourceAction` and `DeleteConversationAction` definition to their own file/module. That module is at the top level at the moment because most actions benefit from being available at the "main widows" level when used with keyboard shortcuts (the widget needs to be in focus for the shortcut to be available, the main window is always in focus). Adding shortcuts for actions is discussed in https://github.com/freedomofpress/securedrop-client/issues/1358

Once we have clarity about which shortcuts we'll add or not add, the actions can be brought down the widget tree when appropriate.

# Test Plan

- [x] Confirm that that sources can be deleted as usual
- [x] Confirm that that conversations can be deleted as usual
- [x] Confirm that the `DeleteSourceAction` definition is not changed significantly (beyond resolving the circular dependency)
- [x] Confirm that the `DeleteSourceAction` tests are not changed significantly (beyond resolving the circular dependency)
- [x] Confirm that the `DeleteConversationAction` definition is not changed significantly (beyond resolving the circular dependency)
- [x] Confirm that the `DeleteConversationAction` tests are not changed significantly (beyond resolving the circular dependency)
- [x] Confirm that the `DownloadConversation` action definition is not changed significantly
- [x] Confirm that the `DownloadConversation` action tests are not changed significantly
- [x] Confirm that the translatable strings haven't changed
- [x] Confirm that CI is green

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

gonzalo-bulnes commented on securedrop_client/gui/widgets.py:
```diff
@@ -3681,7 +3681,7 @@ def __init__(self, source: Source, controller: Controller) -> None:
 class DeleteSourceAction(QAction):
     """Use this action to delete the source record."""
 
-    def __init__(self, source: Source, parent: SourceMenu, controller: Controller) -> None:
```
I introduced this circular dependency when adding type hints. I thought narrower was better, and the circular dependency was hidden by the fact all `SourceMenu` and the actions are defined in the same file.

gonzalo-bulnes commented:
Today's PR @creviera :slightly_smiling_face: ready to review. The actions are a bit different to the dialogs before, because I figure there shouldn't be too many of them at the top level at any one time. (They're are the top level of the GUI tree because they'll likely get keyboard shortcuts, and an application with too many keyboard shortcuts is probably error-prone? If on the other side, we decide not to give those actions a shortcut - that is tracked in #1358 - then we can bring them down into the `source` and `conversation` packages.)

Note that the commits are split, and I think it is easier to review commit by commit in this case.

gonzalo-bulnes commented on securedrop_client/gui/actions.py:
```diff
@@ -0,0 +1,69 @@
+
+        super().__init__(self.text, parent)
+
+        self.confirmation_dialog = DeleteConversationDialog(self.source, self.controller)
```
**Side note**: at the moment, the actual _action_ happens inside the dialogs (that's why they depend on the controller). Eventually, that can be extracted out of the dialogs as follows:

- the action `QAction` creates the dialog
- the action susbcribes tot the `accepted` and `rejected` signals of the dialog
- the actions performs whatever _action_ is required (with help of the `controller`) when the corresponding slots are called

I haven't included that change here in order to keep the PR as simple as possible, but that's what I've in mind.
Once that done, the `actions.py` file would become a proper interface where the GUI and the `controller` interact.

creviera commented on securedrop_client/gui/actions.py:
```diff
@@ -0,0 +1,69 @@
+Over time, this module could become the interface between
+the GUI and the controller.
+
+Copyright (C) 2021 The Freedom of the Press Foundation.
```
see https://github.com/freedomofpress/securedrop-client/issues/1443

gonzalo-bulnes replied:
I removed them like I did in the SecureDrop SDK some time ago. Whichever convention we end up adopting I'm not worried about these files, they'll be easy to track and update as needed.

creviera reviewed (changes requested):
just a quick reminder that there are some conflicts in this PR and that we should avoid adding long licensing headers for now, see https://github.com/freedomofpress/securedrop-client/issues/1443

once the conflicts are resolved, feel free to move this to ready for review!

gonzalo-bulnes commented:
Rebased, updated the test plan (these is one more action - `actions.DownloadConversation`) and removed the copyright and licensing header. This is ready to review @creviera :slightly_smiling_face:

creviera reviewed (approved):
this looks great. nice refactor :)
